### DISCUSSION TITLE
Ability to declare the whole API spec outside of the service's code

### DISCUSSION THREAD

**Discussion by pkucmus** [fastapi/fastapi#6923] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * Or, I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained [minimal, ~reproducible~, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case (not "reproducible" as to my knowledge it's currently not easily available):

```Python
from fastapi import APIRouter


router = APIRouter()


my_route = router.post("/myroute", summary="My Route")
```

```
from fastapi import FastAPI

from my_routes import my_route


app = FastAPI()


@my_route
def my_route_view(my_header: Optional[str] = Header(None),):
    return {"Hello": "World"}
```

### Description

FastAPI does a very good job separating the API spec/contract/definition from the business logic of the application but I'd like to push it a bit further and make that _completely_ possible. It's not _completely_ possible now since one needs to re-declare query params and headers for each view, for example: if those could be declared in the route the problem would be solved (but I'm not sure that would be the proper solution). I saw fastapi-utils with it's class-based views but that's still not it.

Ultimately what I'm looking for is the possibility to hold my entire spec in a Python Package that would contain:

[x] Pydantic definitions
[x] Route definitions
[ ] View query params, headers (cookies, etc) 

That way I can version my "Python defined" OAS API Contract and use it's versions in different services (I know this maybe a weird use case, but I have proxy services that need to have the same API contract on one side but are talking to different business logic providers on the other side).

### The solution you would like

I would like to be able to specify my API definition outside of my business logic

```Python
# definition.py
from fastapi import APIRouter
from pydantic import BaseModel

class MyModel(BaseModel):
    my_field: str

router = APIRouter()

my_route = router.post("/myroute", summary="My Route")
my_endpoint = EndpointDefinition(request_model: MyModel, my_header: Optional[str] = Header(None),)  # maybe?

app = FastAPI()

```

```python
# application.py
from fastapi import FastAPI

from definition import my_route, my_endpoint

app = FastAPI()


@my_route
@my_endpoint
def my_route_view(request_model, headers):
    return "OK"

```

### Describe alternatives you've considered

There's many ways to approach this and in the end one that fits the authors vision for FastAPI would need to be chosen. If it's not something that would fit the general vision then I'm curious to see any ideas people may have. Do you see the value? 
I was thinking about partials, the mentioned class based views (where the class could be the `EndpointDefinition` 

### Environment

* OS: Linux
* FastAPI Version 0.61.2:
* Python version: 3.9

### Additional context

None

Thank you.

**Comment by Mause** [comment]
Is there something wrong with the following?


```python

def example_implementation(authorization: str):
   ...


@app.get('/example')
def example(authorization = Header(...)):
   return example_implementation(authorization)
```

**Comment by pkucmus** [comment]
You are on point - this would be the simplest example - thank you (I think it represents better what I wanted to achieve that my description :D). 
But there is one thing that I'm lacking it that particular example `example_implementation` is the business logic I'd like to "inject" after having my spec defined. When writing the spec (to follow an important to me rule which is "API first") I don't have access to the `example_implementation` function, so what I want to do is "backwards" (from lack of a better word) to this solution.

I want `example` (with it's header, path, etc definitions) to live in a package called `my-spec==1.0.0` and that package would be pip installed by the application that contains the business logic.

One way would be to:

```python
import importlib
import something_like_django_settings

@app.get('/example')
def example(authorization = Header(...)):
   return importlib.import_module(something_like_django_settings.GET_VIEW_PATH)(authorization)
```

But I can't wrap my head around the "settings" part and I just don't like that particular approach. I hope I'm making sense.

**Comment by Mause** [comment]
There are of course many ways to go about this - but here's one that tries to follow the example you gave in your original question:

```python

from inspect import Signature, signature
from typing import Any, Literal, Optional, Type, Union

from fastapi import APIRouter, FastAPI, Header
from fastapi.routing import APIRoute
from fastapi.testclient import TestClient
from makefun import create_function
from pydantic import BaseModel


class DeferredImplementation(BaseModel):
    method: Literal['GET', 'POST', 'PATCH', 'DELETE', 'UPDATE']
    path: str
    response_model: Optional[Type[Any]] = None
    signature: Signature = None

    class Config:
        arbitrary_types_allowed = True

    def __call__(self, function):
        self.signature = signature(function)
        return self

    def implement(self, app: Union[APIRouter, FastAPI]) -> APIRoute:
        def internal(function):
            method = getattr(self.app, self.method.lower())
            route = method(path=self.path, response_model=self.response_model)
            return route(create_function(self.signature, function))

        return internal


@DeferredImplementation(method='GET', path='/example')
def my_endpoint(authorization=Header(...)):
    raise NotImplementedError()


app = FastAPI()


@my_endpoint.implement(app)
async def actual(authorization):
    return authorization


tc = TestClient(app)

assert tc.get('/example', headers={'Authorization': 'hello'}).json() == 'hello'
```

**Comment by Mause** [comment]
(far from production ready code, but it gives you an idea of the concept)

**Comment by pkucmus** [comment]
`makefun` is very new to me. I'll give it a go and I adore there is a way to leverage Pydantic here (probably something that is happening behind the scenes either way, right?). Thank you very much I'll give it a go in the next few days and will report back.

**Comment by pkucmus** [comment, marked as answer]
I was able to achieve what I wanted with `Depends` which I'm happy about as it does not involve any "dark magic".
Basically what I'm doing it:
```python
router = APIRouter()

class MyEndpoint:
    @staticmethod
    def get_router():
        return router.get(
            "/",
            response_model=ResnposeModel,
            responses=responses,
            summary="...",
            response_model_exclude_none=True,  # able to define all of the route parameters
            operation_id="...",
            status_code=200,
        )
    
    def __init__(self, all_of_my_params_models_headers):
        self.all_of_my_params_models_headers = all_of_my_params_models_headers
```

then on an endpoint "view":

```python
@MyEndpoint.get_router()
def my_view(request: MyEndpoint = Depends()):
    return request.all_of_my_params_models_headers
```

I can also prepare an instance of `FastAPI` (with for example OAS `description` and `version`) in the shared package and with that I have a completely defined spec in the package, ready to be used in my application.

The only thing is that I need to lazy include the routes, otherwise those are not included in the OAS spec, I guess it has something to do with how or when the view function is inspected.

I'm good, thanks for the help!

**Comment by Mause** [comment]
Glad to hear you've sorted it out - please feel free to close the issue :)

**Comment by tiangolo** [comment]
Thanks for the help here @Mause ! 👏  🙇 

Thanks for reporting back and closing the issue @pkucmus  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
